torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -2267,21 +2267,24 @@ void InitXlaModuleBindings(py::module m) {
       .def("_get_stablehlo",
            [](const std::vector<at::Tensor>& tensors, const std::string& device,
               const std::vector<std::string>& devices,
-              bool emit_bytecode) -> py::bytes {
-            NoGilSection nogil;
-            EmitMode mode = emit_bytecode ? EmitMode::kStableHloBytecode
-                                          : EmitMode::kStableHloReadable;
-            std::vector<XLATensorPtr> xtensors;
-            if (tensors.empty()) {
-              torch::lazy::BackendDevice backend_device =
-                  GetDeviceOrCurrent(device);
-              xtensors =
-                  XLAGraphExecutor::Get()->GetLiveTensors(&backend_device);
-            } else {
-              xtensors = GetXlaTensors(tensors, /*want_all=*/false);
+              bool emit_bytecode) {
+            std::string hlo_output;
+            {
+              NoGilSection nogil;
+              EmitMode mode = emit_bytecode ? EmitMode::kStableHloBytecode
+                                            : EmitMode::kStableHloReadable;
+              std::vector<XLATensorPtr> xtensors;
+              if (tensors.empty()) {
+                torch::lazy::BackendDevice backend_device =
+                    GetDeviceOrCurrent(device);
+                xtensors =
+                    XLAGraphExecutor::Get()->GetLiveTensors(&backend_device);
+              } else {
+                xtensors = GetXlaTensors(tensors, /*want_all=*/false);
+              }
+              hlo_output = XLAGraphExecutor::Get()->DumpHloComputation(xtensors, mode);
             }
-            return py::bytes(
-                XLAGraphExecutor::Get()->DumpHloComputation(xtensors, mode));
```

Comment:
Is the old behavior a bug? How was the bug introduced without being noticed?

Reply:
I am guessing that the issue is that conversion of std::string -> python str need to be inside of GIL or there could be race condition. 

Probably with new Python version the race condition is now exposed (previous Python version likely had a different order of execution and did not tickle this bug).

Reply:
Yup, this is mentioned in https://github.com/pybind/pybind11/issues/1397. Using any python types requires holding the GIL, otherwise it will lead to a segfault. 

As Han mentioned, previous python versions probably didn't expose this bug.